src/tir/transforms/storage_rewrite.cc

```diff
@@ -1304,6 +1307,18 @@ class VectorTypeAccessChecker : public StmtExprVisitor {
       }
     }
 
+    if (detect_scalar_read_patterns_ && is_buffer_load && indices.size()) {
+      const PrimExpr last_dim_index = indices[indices.size() - 1];
+      if (last_dim_index.dtype().lanes() == 1) {
+        arith::ModularSet me = analyzer_.modular_set(last_dim_index);
+        if (me->coeff > 0) {
+          // When coeff == 0, the index is constant and doesn't need to be recorded since it can
+          // always be rewritten to shuffle.
+          var_info.access_dtype.insert(access_dtype.with_lanes(me->coeff));
```

Comment:
i think we need a data structure that captures two things: 

```c++
struct VarReadInfo {
   DataType access_dtype;
   // maintained as GCD of all coef of access index
   int64_t simd_coeff;
};
```
This is mainly to ensure that we don't end up use a very large vector here

Reply:
access_dtype also contains dtype for write, so the result dtype will be eventually bound by the vectorization lanes of writes

Reply:
the main corner case would be when we do not have writes and only reads.

Reply:
updated